design/dhcp-less.md

```diff
@@ -0,0 +1,391 @@
+assigned to specific clusters)
+
+##### IPAM Scenario 2 - decoupled preprovisioning/provisioning IPPool
+
+An environment where a decoupled configuration is desired for the pre-provisionining
+phase and the provisioned BareMetalHost (e.g BMaaS scenario where end-user network configuration
+differs from the commissioning phase where a different configuration is desired for inspection/cleaning)
+
+## Design Details
+
+`Metal3MachineTemplate` and `Metal3DataTemplate` are used to apply networkData to specific BareMetalHost resources,
+but they are by design coupled to the CAPI Machine lifecycle.
+
+This is a problem for the pre-provisioning use-case since at this point we're preparing the BareMetalHost for
+use, there is not yet any Machine.
+
+To resolve this below we outline a proposal to add two new resources with similar behavior for the pre-provisioning
+phase `Metal3PreProvisioningTemplate` and `Metal3PreProvisioningDataTemplate`
+
+### API overview
+
+The current flow in the provisioning phase is as follows (only the most relevant fields are included for clarity):
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: pool-1
+spec:
+  clusterName: cluster
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3DataTemplate
+metadata:
+  name: data-template
+spec:
+  clusterName: cluster
+  networkData:
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: pool-1
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: machine-template
+spec:
+  template:
+    spec:
+      dataTemplate:
+        name: data-template
+      hostSelector:
+        matchLabels:
+          cluster-role: control-plane
+
+---
+apiVersion: cluster.x-k8s.io/v1beta1
+kind: MachineDeployment
+metadata:
+  name: machine-deployment
+spec:
+  clusterName: cluster
+  replicas: 1
+  template:
+    spec:
+      clusterName: cluster
+      infrastructureRef:
+        apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+        kind: Metal3MachineTemplate
+        name: machine-template
+```
+
+In this flow when a Metal3Machine is provisioned via the `MachineDeployment`, BareMetalHost resources labeled
+`cluster-role: control-plane` will have `networkData` defined with an IP derived from the `pool-1` `IPPool`.
+
+In CAPM3 an IPClaim is created to reserve and IP from the IPPool for each Machine, and an IPAddress resource
+contains the data used for templating of the `networkData`
+
+#### Preprovisioning - Common IPPool
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: pool-1
+spec:
+  clusterName: cluster
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3PreprovisioningDataTemplate
+metadata:
+  name: preprov-data-template
+spec:
+  preprovisioningNetworkData:
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: pool-1
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
```

Comment:
The intention is to emulate the familiar `Metal3MachineTemplate` API but it's *not* coupled to any cluster or Machine - it's just a way of defining a hostSelector so we can match groups of BMH resources to exactly one `Metal3PreprovisioningDataTemplate`

As you have mentioned an alternative would be a new BMH API which consumes the `Metal3PreprovisioningDataTemplate` but that would be inconsistent with the existing `networkData` handling so I worry it could be confusing.

Another alternative could be to create a single new CR which describes both the preprovisioning network config and the hostSelector instead of requiring two CRs?

Reply:
I may be wrong I think it is not confusing because the hostSelector is for scheduling a Bmh for a machine but is not used for the link between a machine and data.

Reply:
@pierrecregut can you help me understand the alternative you are proposing please?

Are you suggesting that I remove `Metal3PreprovisioningTemplate` and add the `hostSelector` to `Metal3PreprovisioningDataTemplate` directly?

Such an approach seems reasonable and you make a good point that in this flow there's no `Machine` so perhaps a resource analogous to `Metal3MachineTemplate` is not required.

Reply:
But the main question is still : why is the dependency inverted ? It is a 1-n dependency. You need a selector if you try to express it from the template side, you don't if you express it from the bmh side and the intent is much clearer for anybody that needs to know if there is preprovisioning template associated to the host. For  example the bmh controller itself that needs to know if it must pause registration or not.

Reply:
@pierrecregut I'm trying to avoid adding any template field to the BMH for the following reasons:

* The BMH/BMO won't consume the template, so it seems inappropriate and potentially confusing to reference it in the BMH spec
* The Scope of this API is outside the BMH - it's layered templating to generate `preprovisioningNetworkDataName` exactly like we already do for `networkData`
* I don't want to couple the BMH to a specific template resource type as it could change in the future (e.g NMState, alternative IPAM solution, or as you've proven with Kanod the templating can be replaced by a downstream alternative templating solution provided we don't couple the BMH and CAPM3 APIs)

So given those concerns I think the options are either a hostSelector, or some kind of Group resource which references a list of BMH resources (I'm trying to avoid this as it would require update on addition of a new BMH unlike the label-based solution)

Reply:
@hardys networkData is associated to the m3machine and the m3machine refers to the template. In fact, because templating is only done at the level of machine and not bmh limits the use we can have of bmh  (use of bmh outside of capi, or with an infra provider that is not capm3). 

Going from a bmh to the templates becomes unnecessarily complex. We have to take care of conflicts when multiple templates apply (and there is no perfect solution).

Point 3 is valid but for me the real solution is that the templater should be generic enough to be able to generate any kind of resource. The ideal would be to be able to specify something like a KRM function but pure go templates answer most of our needs. Problems arise if we want to generate template bound resources like ip address for which we need a custom solution but in fact ip addresses should be attached to the BMH directly not to a given template.

Another solution used to support different ipam is to identify both the template resource name and its kind. As the core BMH controller only really need to understand there is a template and that it should not proceed until it is expanded, we may implement additional template expansion in "sidecar" controllers.

Reply:
@pierrecregut thanks for the feedback!  I think we're approaching this problem from different perspectives, so I'm trying to understand how to make progress:

My goal with this proposal is to define a minimally invasive way to extend the existing APIs, such that Metal3 can operate in DHCP-less scenarios more easily - I want to avoid impacting existing users and find a solution which can be implemented/adopted in a reasonably short timeframe.

Your ideas I think are interesting, but IIUC would imply a far larger API impact - the BMH already has a well defined API for network configuration (Secrets), and with your direction we'd have to add some kind of templating support to that API, move the templating engine from CAPM3 to BMO where I'm not sure it's a good logical fit.

I see two potential paths forward:

1. Can you make specific/actionable suggestions to improve this current proposal?  Are there ways we can address your concerns without massively expanding the scope of this work?
2. If you completely disagree with the direction I've outlined, would you be willing to draft an alternative proposal so the community can discuss and reach consensus on the preferred direction?

Reply:
My suggestion is to remove the selector and either:
* use an annotation as we did in Kanod. If the annotation is part of the contract, the bmh code can act as if it was a paused annotation as long as there is no preprovisioning secret. You can also use either an explicit pause or Zane suggestion but I do not see a strong reason for doing so in that case.
* add an optional field to the BMH CRD (reduced impact for existing deployments): The field should contain a kind and a name. It would be handled as the annotation by an external controller but the BMH controller must know that it should wait for a preprovisioning secret.

In both case the intent is clear, you only need to look at the BMH to understand the behaviour and the impact of both solutions is minor but the second is cleaner and probably easier to expand.

Reply:
Thanks @pierrecregut - I think you're describing two different problems so let me try to summarize:

To prevent the BMH starting inspection before any `preprovisioningNetworkDataName` is set, we could:

1. Use the existing paused annotation and have an external controller remove it when `preprovisioningNetworkDataName` is set (current Kanod approach)
2. Add an explicit `preprovisioningNetworkDataRequired` API to the BMH

Reply:
@zaneb what do you think of the suggestion from @pierrecregut?  

IIUC it would mean something like `preprovisioningNetworkDataTemplateReference` in the BMH spec, which would serve two purposes - act as a flag like `PreprovisioningNetworkDataRequired` and correlate the BMH to a specific template which generates the `PreprovisioningNetworkDataName` Secret.

This would reduce the API complexity compared to my proposal but having a BMH spec field which is not consumed by BMO seems wrong to me from an abstraction/layering perspective.

Using Kind/Name does resolve my biggest concern though which was coupling the BMH API to a specific template type that we may want to change in the future.

Reply:
I don't think we should re-use the pause annotation, because that requires something to come along and remove it again. We should avoid creating a little workflow and aim for declarative - i.e. tell the BMO that it has to wait for the network data.

That could be a separate annotation to instruct the BMO to wait, and if so it would make sense to include info in the annotation data about which object we're waiting for (even though the BMO would only look at the presence of the annotation).

Obviously anything that is in an annotation _could_ be added to the API, but adding the data about which object is going to do the templating seems heavyweight in that case, given that the BMO is not going to do anything with that information. So I think if we're going to add it to the API I'd prefer just a boolean.

And, for completeness, my suggestion was that if a network data Secret is specified, BMO looks at the existence of any content to see when it is ready.

So I'd say I'm not strongly opposed to any of:

1. new annotation
2. boolean in API
3. infer from Secret contents

Reply:
But with a boolean (or empty secret), you do not solve the second problem which is associating a BMH as soon as it is created to a well defined template. With an annotation you could (in fact that is what we do in Kanod with BmhData) but it may be harder to have coexisting templating mechanisms if we need to make it evolve (not necessarily a blocking point).

Reply:
I guess you should read my comments as opinions on the BMH part of the API. I don't really have an opinion on the other side of it :slightly_smiling_face:

Reply:
Thanks @zaneb and @pierrecregut for the discussion - it sounds like an annotation is the approach we can all agree on, and it's the only option under discussion which enables association of templates directly with the BMH.

I also think (provided any new controller in CAPM3 is made optional/configurable) that this solution could allow for alternative templating methods, since the annotation value isn't strictly restricted to a specific resource type.

Reply:
Reworked the proposal to use an annotation instead